specs/experimental/custom-gas-token.md

```diff
@@ -0,0 +1,306 @@
+ensure the security of the system. The source of truth for the configuration of the address of the token is
+in the L1 `SystemConfig` smart contract.
+
+## Constants
+
+| Constant          | Value                           | Description |
+| ----------------- | ------------------------------- | ----------- |
+| `ETHER_TOKEN_ADDRESS` |  `address(0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE)` | Represents ether for gas paying asset |
+| `DEPOSITOR_ACCOUNT` | `address(0xDeaDDEaDDeAdDeAdDEAdDEaddeAddEAdDEAd0001)` | Account with auth permissions on `L1Block` contract |
+| `GAS_PAYING_TOKEN_SLOT` | `bytes32(uint256(keccak256("opstack.gaspayingtoken")) - 1)` | Storage slot that contains the address and decimals of the gas paying token |
+
+## Properties of a Gas Paying Token
+
+The gas paying token MUST satisfy the standard [ERC20](https://eips.ethereum.org/EIPS/eip-20) interface
+and implementation. It MUST be a L1 contract.
+
+The gas paying token cannot:
+
+- Have a fee on transfer
+- Have rebasing logic
+- Have call hooks on transfer
+- Have more than 18 decimals
+- Have out of band methods for modifying balance or allowance
+
+It MUST be enforced on chain that the token has 18 or less decimals to guarantee no losses of precision when
+depositing a token that has a different number of decimals.
+
+## Configuring the Gas Paying Token
+
+The gas paying token is set within the L1 `SystemConfig` smart contract. This allows for easy access to the information
+required to know if an OP Stack chain is configured to use a custom gas paying token. The gas paying token is set
+during initialization and cannot be modified until another initialization. It is assumed that the chain operator
+will not modify the gas paying token address unless they specifically decide to do so and appropriately handle collateralization.
+
+If the gas paying token address is set to the `ETHER_TOKEN_ADDRESS` then the system is configured to use `ether` as the
+native asset on L2. Note that if nothing is stored in state at the `GAS_PAYING_TOKEN_SLOT` then it is considered that
+`ether` is used to pay for gas and getters on smart contracts SHOULD return the `ETHER_TOKEN_ADDRESS`.
+
+When the gas paying token is configured in the `SystemConfig`, it kicks off a special deposit transaction that is from the
+`DEPOSITOR_ACCOUNT` and calls the `L1Block` contract so that it can store the gas paying token address and its number of decimals
+in the L2 system.
+
+![](../static/assets/custom-token-setup.png)
+
+## Contract Modifications
+
+### OptimismPortal
+
+The `OptimismPortal` is updated with a new interface specifically for depositing custom tokens.
+
+#### `depositERC20Transaction`
+
+The `depositERC20Transaction` is useful for sending custom gas tokens to L2. It is broken out into its own interface
+to maintain backwards compatibility, simplify the implementation and prevent footguns when trying to deposit an
+ERC20 native asset into a chain that uses `ether` as its native asset.
```

Comment:
Ok I can expand on the footguns. It would be a large refactor to support the xdm `SendMessage` interface due to the way that the abstractions work. I agree its not ideal

Reply:
Noticed that comment was unresolved—did we expand on this somewhere?

Reply:
This has been addressed. The footgun is simply around being explicit when depositing an ERC20 token, hence the different function name.